File: my-submissions/0698-partition-to-k-equal-sum-subsets/solution.cs
public class Solution {
    public bool CanPartitionKSubsets(int[] nums, int k) {
        bool[] visited = new bool[nums.Length];
        bool result = false;
        int targetSum = 0;
        foreach(int i in nums) targetSum += i;
        if(targetSum % k == 0) targetSum /= k;
        else return false;
        Array.Sort(nums);
        Array.Reverse(nums);
        if (nums[0] > targetSum) return false;
        bool backtrack(int start, int currSum, int remainingSets)
        {
            if(remainingSets == 0) return true;
            if(currSum == targetSum) return backtrack(0, 0, remainingSets - 1);
            for(int j = start; j < nums.Length; j++)
            {
                if(visited[j] || currSum + nums[j] > targetSum) continue;
                visited[j] = true;
                if(backtrack(j + 1, currSum + nums[j], remainingSets))
                {
                    return true;
                }
                visited[j] = false;
                if(currSum == 0) break;
            }
            return false;
        }
        return backtrack(0, 0, k);
    }
}

File: my-submissions/0698-partition-to-k-equal-sum-subsets/solution.java
class Solution {
    public boolean canPartitionKSubsets(int[] nums, int k) {
        boolean dp[] = new boolean[1 << nums.length];
        int maskSum[] = new int[1 << nums.length];
        dp[0] = true;
        int sum = 0, target = 0;
        
        for (int item : nums) sum += item;
        if(sum % k != 0) return false;
        target = sum / k;
        for (int num : nums) if (num > target) return false;

        for (int mask = 0; mask < (1 << nums.length); mask++)
        {
            if(!dp[mask]) continue;
            for(int i = 0; i < nums.length; i++)
            {
                if((mask & (1 << i)) != 0) continue;
                int newMask = mask | (1 << i);
                if(dp[newMask]) continue;
                if(maskSum[mask] % target + nums[i] <= target)
                {
                    maskSum[newMask] = maskSum[mask] + nums[i];
                    dp[newMask] = true;
                }
            }
        }

        return dp[(1 << nums.length) - 1];
    }
}
